src/mapgen_valleys.cpp

```diff
@@ -0,0 +1,1013 @@
+	noise_inter_valley_slope->perlinMap2D(x, z);
+	noise_rivers->perlinMap2D(x, z);
+	noise_terrain_height->perlinMap2D(x, z);
+	noise_valley_depth->perlinMap2D(x, z);
+	noise_valley_profile->perlinMap2D(x, z);
+
+	if (fast_terrain) {
+		// Make this 2D for speed, if requested.
+		noise_inter_valley_fill->perlinMap2D(x, z);
+
+		if (cliff_terrain)
+			noise_cliffs->perlinMap2D(x, z);
+		if (rugged_terrain)
+			noise_corr->perlinMap2D(x, z);
+	} else {
+		noise_inter_valley_fill->perlinMap3D(x, y, z);
+	}
+
+	if (flags & MG_CAVES) {
+		noise_simple_caves_1->perlinMap3D(x, y, z);
+		noise_simple_caves_2->perlinMap3D(x, y, z);
+	}
+
+	//mapgen_profiler->avg("noisemaps", tcn.stop() / 1000.f);
+
+	for (s32 index = 0; index < csize.X * csize.Z; index++) {
+		noise_heat->result[index] += noise_heat_blend->result[index];
+		noise_heat->result[index] += temperature_adjust;
+		noise_humidity->result[index] += noise_humidity_blend->result[index];
+	}
+
+	u32 index = 0;
+	for (s16 zi = node_min.Z; zi <= node_max.Z; zi++)
+	for (s16 xi = node_min.X; xi <= node_max.X; xi++, index++) {
+		// The parameters that we actually need to generate terrain
+		//  are passed by address (and the return value).
+		float mount = terrainLevelFromNoise(xi, zi,
+				noise_terrain_height->result[index],
+				// River noise is replaced with base terrain, which
+				// is basically the height of the water table.
+				&noise_rivers->result[index],
+				// Valley depth noise is replaced with the valley
+				// number that represents the height of terrain
+				// over rivers and is used to determine about
+				// how close a river is for humidity calculation.
+				&noise_valley_depth->result[index],
+				noise_valley_profile->result[index],
+				// Slope noise is replaced by the calculated slope
+				// which is used to get terrain height in the slow
+				// method, to create sharper mountains.
+				&noise_inter_valley_slope->result[index],
+				noise_inter_valley_fill->result[index],
+				noise_cliffs->result[index],
+				noise_corr->result[index]);
+
+		noise_terrain_height->result[index] = mount;
+
+		if (fast_terrain) {
+			// Assign the humidity adjusted by water proximity.
+			// I can't think of a reason why a mod would expect base humidity
+			//  from noise or at any altitude other than ground level.
+			noise_humidity->result[index] = humidityByTerrain(
+					noise_humidity->result[index],
+					mount,
+					noise_rivers->result[index],
+					noise_valley_depth->result[index]);
+
+			// Assign the heat adjusted by altitude. See humidity, above.
+			if (use_altitude_chill && mount > 0.f)
+				noise_heat->result[index] *= pow(0.5f, (mount - altitude_chill / 3.f) / altitude_chill);
+		}
+	}
+
+	heatmap = noise_heat->result;
+	humidmap = noise_humidity->result;
+}
+
+
+// This ugly function keeps us from having to maintain two similar sets of
+//  complicated code to determine ground level.
+float MapgenValleys::terrainLevelFromNoise(
+		s16 x,
```

Comment:
i think you should group all those parameters into a structure passed by reference if possible to reduce memory copies